be/src/exec/vectorized/aggregate/agg_hash_map.h

```diff
@@ -119,64 +119,90 @@ struct AggHashMapWithOneNumberKey : public AggHashMapWithKey<HashMap> {
     AggDataPtr get_null_key_data() { return nullptr; }
 
     // prefetch branch better performance in case with larger hash tables
-    template <typename Func>
-    void compute_agg_prefetch(ColumnType* column, Buffer<AggDataPtr>* agg_states, Func&& allocate_func) {
+    template <typename Func, bool allocate_and_compute_state, bool compute_not_founds>
+    void compute_agg_prefetch(ColumnType* column, Buffer<AggDataPtr>* agg_states, Func&& allocate_func,
+                              std::vector<uint8_t>* not_founds) {
         AGG_HASH_MAP_PRECOMPUTE_HASH_VALUES(column, AGG_HASH_MAP_DEFAULT_PREFETCH_DIST);
         for (size_t i = 0; i < column_size; i++) {
             AGG_HASH_MAP_PREFETCH_HASH_VALUE();
 
             FieldType key = column->get_data()[i];
-            auto iter = this->hash_map.lazy_emplace_with_hash(key, hash_values[i], [&](const auto& ctor) {
-                AggDataPtr pv = allocate_func(key);
-                ctor(key, pv);
-            });
-            (*agg_states)[i] = iter->second;
+
+            if constexpr (allocate_and_compute_state && compute_not_founds) {
+                auto iter = this->hash_map.lazy_emplace_with_hash(key, hash_values[i], [&](const auto& ctor) {
+                    (*not_founds)[i] = 1;
+                    AggDataPtr pv = allocate_func(key);
+                    ctor(key, pv);
+                });
+                (*agg_states)[i] = iter->second;
+            } else if constexpr (compute_not_founds) {
+                DCHECK(not_founds);
+                if (auto iter = this->hash_map.find(key); iter != this->hash_map.end()) {
+                    (*agg_states)[i] = iter->second;
+                } else {
+                    (*not_founds)[i] = 1;
+                }
+            } else if constexpr (allocate_and_compute_state) {
+                auto iter = this->hash_map.lazy_emplace_with_hash(key, hash_values[i], [&](const auto& ctor) {
+                    AggDataPtr pv = allocate_func(key);
+                    ctor(key, pv);
+                });
+                (*agg_states)[i] = iter->second;
+            }
         }
     }
 
     // prefetch branch better performance in case with small hash tables
-    template <typename Func>
-    void compute_agg_noprefetch(ColumnType* column, Buffer<AggDataPtr>* agg_states, Func&& allocate_func) {
+    template <typename Func, bool allocate_and_compute_state, bool compute_not_founds>
+    void compute_agg_noprefetch(ColumnType* column, Buffer<AggDataPtr>* agg_states, Func&& allocate_func,
```

Comment:
as above, explicitly, three different purposes split into single function would make the code more clear and succinct.